rustworkx-core/src/coloring.rs

```diff
@@ -141,21 +176,190 @@ where
     Ok(colors)
 }
 
+/// Data associated to nodes for the greedy coloring algorithm
+/// using the "saturation first" strategy: always picking the node that
+/// has the largest number of different colors already assigned to its
+/// neighbors, and, in case of a tie, the node that has the largest number
+/// of uncolored neighbors.
+#[derive(Clone, Eq, PartialEq)]
+struct SaturationStrategyData {
+    // degree of a node: number of neighbors without color
+    degree: usize,
+    // saturation degree of a node: number of colors used by neighbors
+    saturation: usize,
+}
+
+impl Ord for SaturationStrategyData {
+    fn cmp(&self, other: &SaturationStrategyData) -> Ordering {
+        self.saturation
+            .cmp(&other.saturation)
+            .then_with(|| self.degree.cmp(&other.degree))
+    }
+}
+
+impl PartialOrd for SaturationStrategyData {
+    fn partial_cmp(&self, other: &SaturationStrategyData) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+fn inner_greedy_node_color_strategy_saturation<G, F, E>(
+    graph: G,
+    mut preset_color_fn: F,
+) -> Result<DictMap<G::NodeId, usize>, E>
+where
+    G: NodeCount + IntoNodeIdentifiers + IntoEdges + NodeIndexable,
+    G::NodeId: Hash + Eq + Send + Sync,
+    F: FnMut(G::NodeId) -> Result<Option<usize>, E>,
+{
+    let mut colors: DictMap<G::NodeId, usize> = DictMap::with_capacity(graph.node_count());
+
+    let mut nbd_colors: HashMap<G::NodeId, HashSet<usize>> =
+        HashMap::with_capacity(graph.node_count());
+    for k in graph.node_identifiers() {
+        nbd_colors.insert(k, HashSet::new());
+    }
```

Comment:
Not that it really matters since you're pre-allocating, but you could do this as a map collect iterator:
```suggestion
    let mut nbd_colors: HashMap<G::NodeId, HashSet<usize>> = graph.node_identifiers().map(|k| {
        (k, HashSet::new())
    })
    .collect();
```